src/DefaultBuilder/src/WebHostEnvironment.cs

```diff
@@ -106,13 +105,35 @@ public void ResolveFileProviders(IConfiguration configuration)
             }
         }
 
+        private string ResolveWebRootPath(string webRootPath, string contentRootPath)
+        {
+            if (string.IsNullOrEmpty(webRootPath))
+            {
+                return Path.GetFullPath(contentRootPath);
+            }
+
+            if (Path.IsPathRooted(webRootPath))
+            {
+                return webRootPath;
+            }
+
+            return Path.Combine(Path.GetFullPath(contentRootPath), webRootPath);
+        }
+
         public string ApplicationName { get; set; }
         public string EnvironmentName { get; set; }
 
         public IFileProvider ContentRootFileProvider { get; set; }
         public string ContentRootPath { get; set; }
 
         public IFileProvider WebRootFileProvider { get; set; }
-        public string WebRootPath { get; set; }
+
+        // Mimick the setting used in HostingEnvironment
+        private string _webRootPath = default!;
+        public string WebRootPath
+        {
+            get => _webRootPath;
+            set => _webRootPath = ResolveWebRootPath(value, ContentRootPath);
```

Comment:
What is this mimicking exactly? Does any other `IWebHostEnvironment` not set the `WebRootPath` to exactly what it's assigned to?

Reply:
I think we do need to update WebRootPath, but we should probably only do it when we call `WebApplicationBuilder.Build()`. I haven't looked into it super deeply, so I could probably be convinced otherwise.

It looks like the ContentRootPath can also be relative to the executing assembly. When I use the generic host and write something like this:

```C#
public static IHostBuilder CreateHostBuilder(string[] args) =>
    Host.CreateDefaultBuilder(args)
    .ConfigureWebHostDefaults(webBuilder =>
    {
        webBuilder.UseContentRoot("www");
        webBuilder.UseWebRoot("www");
        webBuilder.UseStartup<Startup>();
    });
```

I get this output:

```
C:\dev\halter73\swagger\WebApi webroot ↑1  ~4 ?1 ❯ dotnet run
...
info: Microsoft.Hosting.Lifetime[0]
      Content root path: C:\dev\halter73\swagger\WebApi\bin\Debug\net6.0\www
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/2 GET https://localhost:5001/www.html - -
info: Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware[2]
      Sending file. Request path: '/www.html'. Physical path: 'C:\dev\halter73\swagger\WebApi\bin\Debug\net6.0\www\www\www.html'
...
```

This seems to indicate to me that the content root is relative to `Path.GetDirectoryName(Assembly.GetEntryAssembly().Location)` (or maybe `GetExecutingAssembly()`). @Tratcher do you know where this logic is?

Reply:
I think the logic for setting the `ContentRootPath` is located here:

https://github.com/dotnet/aspnetcore/blob/2383d442acf041d65ffb90e4f1cd9a5bbafc01e3/src/Hosting/Hosting/src/WebHostBuilder.cs#L266-L269

The call made to the `Initialize` combines the `ContentRootPath` when setting the `WebRootPath`.

https://github.com/dotnet/aspnetcore/blob/0ae054093c772562ab6c71a042c5d018aa715309/src/Hosting/Hosting/src/Internal/HostingEnvironmentExtensions.cs#L87-L99

Reply:
> I think we do need to update WebRootPath, but we should probably only do it when we call WebApplicationBuilder.Build(). I haven't looked into it super deeply, so I could probably be convinced otherwise.

I'm assuming you refer to [this](https://github.com/dotnet/aspnetcore/blob/2383d442acf041d65ffb90e4f1cd9a5bbafc01e3/src/DefaultBuilder/src/WebApplicationBuilder.cs#L106). With that approach, we'd end up doing the full-path resolution in:

https://github.com/dotnet/aspnetcore/blob/2383d442acf041d65ffb90e4f1cd9a5bbafc01e3/src/DefaultBuilder/src/ConfigureWebHostBuilder.cs#L74

Since the invocation path is WebApplication.Build => ConfigureWebHost => ApplySettings => UseSetting.

https://github.com/dotnet/aspnetcore/blob/2383d442acf041d65ffb90e4f1cd9a5bbafc01e3/src/DefaultBuilder/src/ConfigureWebHostBuilder.cs#L97-L100

But doing it here wouldn't account for the scenario where we want to support users setting the `WebRootPath` directly via `builder.Environment.WebRootPath = "foo"`.

Reply:
Could we call ResolveFileProviders at the start of ApplyEnvironmentSettings to handle that?

Reply:
And then ResolveFileProviders would include the ResolveWebRootPath/ResolveContentRootPath logic.